resolver/addressable_resolver.go

```diff
@@ -211,3 +211,109 @@ func (r *URIResolver) URIFromObjectReference(ctx context.Context, ref *corev1.Ob
+		deprecatedObjectReference = &corev1.ObjectReference{
+			Kind:       dest.DeprecatedKind,
+			APIVersion: dest.DeprecatedAPIVersion,
+			Name:       dest.DeprecatedName,
+			Namespace:  dest.DeprecatedNamespace,
+		}
+	}
+	var addr *duckv1.Addressable
+	if dest.Ref != nil && deprecatedObjectReference != nil {
+		return addr, errors.New("ref and [apiVersion, kind, name] can't be both present")
+	}
+	var ref *corev1.ObjectReference
+	if dest.Ref != nil {
+		ref = dest.Ref
+	} else {
+		ref = deprecatedObjectReference
+	}
+
+	if ref != nil {
+		addr, err := r.AddressableFromObjectReference(ctx, dest, parent)
+		if err != nil {
+			return addr, err
+		}
+		return addr, nil
+	}
+
+	return addr, errors.New("destination missing Ref, [apiVersion, kind, name] and URI, expected at least one")
+}
+
+// AddressableFromDestination resolves a v1.Destination into a addressable.
+func (r *URIResolver) AddressableFromDestinationV1(ctx context.Context, dest duckv1.Destination, parent interface{}) (*duckv1.Addressable, error) {
+	var addr *duckv1.Addressable
+	if dest.Ref != nil {
+		addr, err := r.AddressableFromKReference(ctx, dest, parent)
+		if err != nil {
+			return nil, err
+		}
+		return addr, nil
+	}
+
+	return addr, errors.New("destination missing Ref, [apiVersion, kind, name] and URI, expected at least one")
+}
+
+func (r *URIResolver) AddressableFromKReference(ctx context.Context, dest duckv1.Destination, parent interface{}) (*duckv1.Addressable, error) {
+	kRef := dest.Ref
+	ref := &corev1.ObjectReference{Name: kRef.Name, Namespace: kRef.Namespace, APIVersion: kRef.APIVersion, Kind: kRef.Kind}
+
+	gvr, _ := meta.UnsafeGuessKindToResource(ref.GroupVersionKind())
+	if err := r.tracker.TrackReference(tracker.Reference{
+		APIVersion: ref.APIVersion,
+		Kind:       ref.Kind,
+		Namespace:  ref.Namespace,
+		Name:       ref.Name,
+	}, parent); err != nil {
+		return nil, fmt.Errorf("failed to track reference %s %s/%s: %w", gvr.String(), ref.Namespace, ref.Name, err)
+	}
+
+	lister, err := r.listerFactory(gvr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get lister for %s: %w", gvr.String(), err)
+	}
+
+	obj, err := lister.ByNamespace(ref.Namespace).Get(ref.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get object %s/%s: %w", ref.Namespace, ref.Name, err)
+	}
+
+	addressable, ok := obj.(*duckv1.AddressableType)
+	if !ok {
+		return nil, apierrs.NewBadRequest(fmt.Sprintf("%+v (%T) is not an AddressableType", ref, ref))
+	}
+	if addressable.Status.Addresses != nil && len(addressable.Status.Addresses) != 0 {
+		if dest.Ref.Address == nil || *dest.Ref.Address == "" {
+			return &addressable.Status.Addresses[0], nil
+		}
+		for _, addr := range addressable.Status.Addresses {
+			if addr.Name == dest.Ref.Address {
+				return &addr, nil
+			}
+		}
+		return nil, apierrs.NewBadRequest(fmt.Sprintf("address with name %s cannot be found/resolved for %+v", *addressable.Status.Address.Name, ref))
+	}
```

Comment:
Is this the right flow?

Reply:
also, I undid the changes in `URIFromObjectReference` which allowed it to use `AddressableFromObjectReference`